### PR TITLE
Feature/#70 autocomplete in cli

### PR DESCRIPTION
**Pull request by frankschmitt**
Add a command for generating auto-completion scripts for Bash and Zsh. Implements #70

### CODE REVIEW

File: buildSrc/src/main/kotlin/Versions.kt [hunk 1]
```diff
@@ -10,7 +10,8 @@ object Versions {
     const val testcontainers = "1.16.2"
     const val strikt = "0.32.0"
     const val mockk = "1.9.3"
-    const val clikt = "2.3.0"
+    //const val clikt = "2.3.0"
+    const val clikt = "3.5.0"
```
Comment (Lysoun) [inline buildSrc/src/main/kotlin/Versions.kt:13]: You can remove that comment

File: modules/cli/build.gradle.kts [hunk 2]
```diff
@@ -17,7 +17,8 @@ dependencies {
     implementation(project(":modules:driver-mongo-mapping:driver-mongo-mapping-serialization"))
     implementation(project(":modules:driver-jdbc"))
 
-    implementation("com.github.ajalt:clikt:${Versions.clikt}")
+    implementation("com.github.ajalt.clikt:clikt:${Versions.clikt}")
+    //implementation("com.github.ajalt:clikt:${Versions.clikt}")
```
Comment (Lysoun) [inline modules/cli/build.gradle.kts]: You could remove the comment

File: modules/cli/src/test/kotlin/datamaintain/cli/app/AppTest.kt [hunk 3]
```diff
@@ -201,5 +205,18 @@ internal class AppTest : BaseCliTest() {
...
+        @Nested
+        inner class GenerateCompletion {
+            private val logger = LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME) as Logger
+            private val testAppender = TestAppender()
+
+            @BeforeEach
+            fun setupLogger() {
+                logger.addAppender(testAppender)
+                testAppender.start()
+            }
```
Comment (Lysoun) [inline modules/cli/src/test/kotlin/datamaintain/cli/app/AppTest.kt]: This test class does nothing, doesn't it?

File: modules/cli/src/main/kotlin/datamaintain/cli/completion/RebuildAutoCompletionScripts.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import java.io.File
+import java.io.IOException
+
+data class ProcessResult(val exitValue: Int, val output: String)
+
+fun generateAutoCompletionScripts(outputDir: String) {
+    listOf("bash", "zsh").forEach {
+        val result = execAppInSubprocess(listOf("--generate-completion", it),
+                                 outputDir + "/" + it + "-autocomplete.sh")
+        //File(outputDir + "/" + it + "-autocomplete.sh").writeText(result.output)
```
Comment (Lysoun) [inline modules/cli/src/main/kotlin/datamaintain/cli/completion/RebuildAutoCompletionScripts.kt]: This commented line could be deleted

**Review by Lysoun — CHANGES_REQUESTED** [review]
That looks very promising! Did you try using the autocompletion? I sourced the `zsh-automplete.sh` file and tried using the CLI but couldn't find any completion. Maybe I used it wrongly? You can build the CLI using the following command in the CLI module
```gradle installDist```
and then you will find it in `modules/cli/build/install/clin/bin`

In my opinion, the scripts should not be committed in the repository. They should be included and sourced in our docker packages and in the CLI file that is uploading on every release but if it sounds difficult to you, I can do it. 

I'd prefer the option to build the CLI not to be available in `App` to avoid displaying it in the help and making it a valid option for our users to call. Could you change the implementation and make it read a global variable, that will have to be set during our releases?

File: buildSrc/src/main/kotlin/Versions.kt (hunk 1 above)
Comment (frankschmitt) [reply buildSrc/src/main/kotlin/Versions.kt:13]: Done.

File: modules/cli/build.gradle.kts (hunk 2 above)
Comment (frankschmitt) [reply modules/cli/build.gradle.kts]: Done.

File: modules/cli/src/test/kotlin/datamaintain/cli/app/AppTest.kt (hunk 3 above)
Comment (frankschmitt) [reply modules/cli/src/test/kotlin/datamaintain/cli/app/AppTest.kt]: Doh! Seems I forgot to remove that class after moving tests to a separate sub-package, sorry. I've removed it.

File: modules/cli/src/main/kotlin/datamaintain/cli/completion/RebuildAutoCompletionScripts.kt (hunk 4 above)
Comment (frankschmitt) [reply modules/cli/src/main/kotlin/datamaintain/cli/completion/RebuildAutoCompletionScripts.kt]: Done.

**Comment by frankschmitt** [comment]
> That looks very promising! Did you try using the autocompletion? I sourced the `zsh-automplete.sh` file and tried using the CLI but couldn't find any completion. Maybe I used it wrongly? 

I guess that's because currently, the autocompletion uses the (very non-specific) name "app". See my comment at https://github.com/4sh/datamaintain/issues/70#issuecomment-1278163455 .

After generating the zsh script and sourcing it, auto-completion for the "app" command should work like this

<pre><font color="#EF2929"><b>➜  </b></font><font color="#34E2E2"><b>datamaintain</b></font> <font color="#729FCF"><b>git:(</b></font><font color="#EF2929"><b>feature/#70-autocomplete-in-cli</b></font><font color="#729FCF"><b>) </b></font><font color="#FCE94F"><b>✗</b></font> source bla.sh
<font color="#8AE234"><b>➜  </b></font><font color="#34E2E2"><b>datamaintain</b></font> <font color="#729FCF"><b>git:(</b></font><font color="#EF2929"><b>feature/#70-autocomplete-in-cli</b></font><font color="#729FCF"><b>) </b></font><font color="#FCE94F"><b>✗</b></font> app
generate-completion      list                     mark-script-as-executed  update-db              
</pre>

> You can build the CLI using the following command in the CLI module `gradle installDist` and then you will find it in `modules/cli/build/install/clin/bin`

Nice! I was wondering how to run the CLI, actually. Thanks!

> In my opinion, the scripts should not be committed in the repository. They should be included and sourced in our docker packages and in the CLI file that is uploading on every release but if it sounds difficult to you, I can do it.

I'll give it at shot.

> I'd prefer the option to build the CLI not to be available in `App` to avoid displaying it in the help and making it a valid option for our users to call. Could you change the implementation and make it read a global variable, that will have to be set during our releases?

Sure (I'm not yet sure whether I need to exclude the class somehow from compilation altogether or whether wrapping the call that registers the command is sufficient, but I'll figure it out).

**Comment by frankschmitt** [comment]
@Lysoun I've adapted the code; in principle, it should now work as required:
- the task generate-completion is no longer available in App (and therefore, end users won't see it)
- the gradle task for generating the completion scripts is still available (so devs can create it)
- the completion scripts for bash and zsh were removed; I've added an empty .gitkeep file to the directory docs/auto-completion because it's required for the gradle task
- the docker build scripts now call the gradle task for generating auto completion

However, a couple of things still need to be done / clarified:
- app name for auto-completion. This is currently still "app"; I could create a sub-class of App with a more sensible name (e.g. Datamaintain) and use that for generating the scripts. However, then we would have mismatch between the name used for auto-completion and the "real" app name.
  I'm not sure how we should handle this. One possible approach would be to rename App.kt to Datamaintain.kt (or similar) - do you think that's sensible?
- regarding the Docker images: I'm not sure how that's supposed to work. Should the auto-completion scripts be deployed in the Docker container? Or do you expect users to create an alias for running datamaintain via docker
  (something like alias dm="docker run --it --rm datamaintain:<release>-mongo") and use the auto-completion on the Docker host?
- regarding the CLI file: how is this packaged? As far as I can see, cli / cli.bat is auto-generated by Gradle. Should the auto-completion scripts be bundled insided cli.tar?

**Comment by Lysoun** [comment]
> However, a couple of things still need to be done / clarified:
> 
> * app name for auto-completion. This is currently still "app"; I could create a sub-class of App with a more sensible name (e.g. Datamaintain) and use that for generating the scripts. However, then we would have mismatch between the name used for auto-completion and the "real" app name.
>   I'm not sure how we should handle this. One possible approach would be to rename App.kt to Datamaintain.kt (or similar) - do you think that's sensible?

I think you could just rename the class `App` without renaming the file, though it's named quite badly; it's not clear that the CLI's main is in that file.

> * regarding the Docker images: I'm not sure how that's supposed to work. Should the auto-completion scripts be deployed in the Docker container? Or do you expect users to create an alias for running datamaintain via docker
>   (something like alias dm="docker run --it --rm datamaintain:-mongo") and use the auto-completion on the Docker host?

Actually we already have docker packages that run Datamaintain so I was not thinking of an alias. I was thinking of deploying the auto-completion scripts and put them in the `.bashrc` of the dockers packages so that you can take advantage from it when you are inside the container. All the Dockerfiles used to build our images are in `modules/cli/Dockerfile`.

> * regarding the CLI file: how is this packaged? As far as I can see, cli / cli.bat is auto-generated by Gradle. Should the auto-completion scripts be bundled insided cli.tar?

I would find it nice, with a little README to explain how to enable autocompletion (just give the set of commands)

**Comment by frankschmitt** [comment]
Sounds good, I‘ll
- rename App to something more unique
- adapt the Dockerfiles / build process
- adapt the build process for cli.tar

**Comment by frankschmitt** [comment]
@Lysoun I've implemented the changes. Auto-completion now "listens" to datamaintain, and the scripts are included in the release and the Docker images. Looking forward to your review :-)

File: docs/auto-completion/README.md [hunk 5]
```diff
@@ -0,0 +1,16 @@
+# SYNOPSIS
+Auto-completion scripts to be used with bash and zsh.
+
+# BUILD
+Run 
+```./gradlew rebuildCompletion```
```
Comment (Lysoun) [inline docs/auto-completion/README.md]:
```suggestion
```./gradlew rebuildAutocompletion```
```

File: modules/cli/Dockerfile/mongo/4/Dockerfile [hunk 6]
```diff
@@ -30,4 +30,8 @@ RUN mkdir /code/datamaintain \
     && rm /code/datamaintain-cli.tar \
     && rm /code/datamaintain/bin/cli.bat
 
+# Setup autocompletion
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
 ENTRYPOINT ["/code/datamaintain/bin/cli"]
```
Comment (Lysoun) [inline modules/cli/Dockerfile/mongo/4/Dockerfile]:
```suggestion
ENTRYPOINT ["datamaintain"]
```

File: modules/cli/Dockerfile/mongo/5/Dockerfile [hunk 7]
```diff
@@ -27,4 +27,8 @@ RUN mkdir /code/datamaintain \
     && rm /code/datamaintain-cli.tar \
     && rm /code/datamaintain/bin/cli.bat
 
+# Setup autocompletion
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
 ENTRYPOINT ["/code/datamaintain/bin/cli"]
```
Comment (Lysoun) [inline modules/cli/Dockerfile/mongo/5/Dockerfile]:
```suggestion
ENTRYPOINT ["datamaintain"]
```

File: modules/cli/Dockerfile/mongo/4/Dockerfile [hunk 8]
```diff
@@ -30,4 +30,8 @@ RUN mkdir /code/datamaintain \
     && rm /code/datamaintain-cli.tar \
     && rm /code/datamaintain/bin/cli.bat
 
+# Setup autocompletion
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
```
Comment (Lysoun) [inline modules/cli/Dockerfile/mongo/4/Dockerfile]:
```suggestion
RUN alias datamaintain=/code/datamaintain/bin/cli
```

File: modules/cli/Dockerfile/mongo/5/Dockerfile [hunk 9]
```diff
@@ -27,4 +27,8 @@ RUN mkdir /code/datamaintain \
     && rm /code/datamaintain-cli.tar \
     && rm /code/datamaintain/bin/cli.bat
 
+# Setup autocompletion
+RUN echo "source /code/datamaintain/bash-autocomplete.sh" >> ~/.bashrc \
+    && echo "source /code/datamaintain/zsh-autocomplete.sh" >> ~/.zshrc
+
```
Comment (Lysoun) [inline modules/cli/Dockerfile/mongo/5/Dockerfile]:
```suggestion
RUN alias datamaintain=/code/datamaintain/bin/cli
```

File: docs/auto-completion/README.md [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+in the project's root directory. Afterwards, you'll find the generated scripts in this directory.
+
+# USAGE
+Source the script for your shell, e.g. for Bash:
+
+```source bash-autocomplete.sh```
```
Comment (Lysoun) [inline docs/auto-completion/README.md:12]:
```suggestion
```source bash-autocomplete.sh```
Define an alias to the Datamaintain CLI:
```alias datamaintain=/your/path/cli```
```

**Review by Lysoun — APPROVED** [review]
Thank you so much for your work, I'm approving your PR!

**Comment by Lysoun** [comment]
I'm gonna merge it in branch 2.1 after fixing some conflicts

**Pull request merged**